File: web_pdf_generator.py
import csv
import io
from js import document, Uint8Array, Blob, URL
from pyodide.ffi import create_proxy
import math
from fpdf import FPDF
import qrcode


# Global variables to store configuration
config = {}
csv_data = None

# Default colors for common card categories
DEFAULT_COLORS = {
    "song": {"bg": "#8cbeb2", "text": "#192623"},
    "video": {"bg": "#f3b562", "text": "#33230e"},
    "article": {"bg": "#f06060", "text": "#320e0e"},
    "painting": {"bg": "#f2ebbf", "text": "#333126"},
}


def rgb_from_hex(hex_color):
    """Convert hex color to RGB tuple."""
    hex_color = hex_color.lstrip("#")
    return tuple(int(hex_color[i : i + 2], 16) for i in (0, 2, 4))


def show_error(message):
    """Display an error message to the user."""
    from js import window

    window.showAlert(message, "error")


def show_success(message):
    """Display a success message to the user."""
    from js import window

    window.showAlert(message, "success")


def show_status(message):
    """Display a status message to the user."""
    from js import window

    window.showAlert(message, "status")


def generate_card_type_colors(card_types):
    """Generate color pickers for each unique card type found in the CSV."""
    container = document.getElementById("dynamic-card-types")
    container.innerHTML = ""  # Clear existing color pickers

    # Hide the "no CSV" message
    no_csv_msg = document.getElementById("no-csv-message")
    if no_csv_msg:
        no_csv_msg.classList.add("hidden")

    # Generate color picker for each card type
    for card_type in sorted(card_types):
        # Get default colors or generate random ones
        default_colors = DEFAULT_COLORS.get(
            card_type, {"bg": "#cccccc", "text": "#000000"}
        )

        # Create the HTML structure for background color
        bg_div = document.createElement("div")

        bg_label = document.createElement("span")
        bg_label.className = "font-medium text-slate-700 dark:text-slate-300"
        bg_label.textContent = f"{card_type.capitalize()} BG"
        bg_div.appendChild(bg_label)

        bg_input = document.createElement("input")
        bg_input.type = "color"
        bg_input.className = "ring-2 ring-slate-200 dark:ring-slate-600 rounded-full"
        bg_input.id = f"{card_type}-bg"
        bg_input.value = default_colors["bg"]
        bg_div.appendChild(bg_input)

        container.appendChild(bg_div)

        # Create the HTML structure for text color
        text_div = document.createElement("div")

        text_label = document.createElement("span")
        text_label.className = "font-medium text-slate-700 dark:text-slate-300"
        text_label.textContent = f"{card_type.capitalize()} Text"
        text_div.appendChild(text_label)

        text_input = document.createElement("input")
        text_input.type = "color"
        text_input.className = "ring-2 ring-slate-200 dark:ring-slate-600 rounded-full"
        text_input.id = f"{card_type}-text"
        text_input.value = default_colors["text"]
        text_div.appendChild(text_input)

        container.appendChild(text_div)


def validate_csv_data(data):
    """Validate that the CSV data has the required columns and valid values."""
    required_columns = ["type", "url", "top", "center", "bottom"]

    if not data:
        return False, "CSV data is empty"

    # Check if all required columns exist
    missing_columns = [col for col in required_columns if col not in data[0].keys()]
    if missing_columns:
        return False, f"Missing required columns: {', '.join(missing_columns)}"

    # Check required fields are not empty (center can be empty)
    for idx, row in enumerate(data, start=1):
        if not str(row.get("type", "")).strip():
            return False, f"Type column cannot contain empty values (row {idx})"
        if not str(row.get("url", "")).strip():
            return False, f"URL column cannot contain empty values (row {idx})"
        if not str(row.get("top", "")).strip():
            return False, f"Top column cannot contain empty values (row {idx})"
        if not str(row.get("bottom", "")).strip():
            return False, f"Bottom column cannot contain empty values (row {idx})"

    return True, "CSV data is valid"


def validate_config(config):
    """Validate that the configuration is valid (basic server-side check)."""
    # Note: Main validation is done on the frontend in JavaScript
    # This is a backup validation in case JavaScript is disabled or bypassed

    # Check if all values are positive
    for key, value in config.items():
        if key.startswith("CARD_COLORS"):
            continue
        if value < 0:
            return False, f"Configuration value '{key}' must be positive"

    return True, "Configuration is valid"


async def load_csv_file(event):
    """Load and validate the CSV file."""
    global csv_data

    file_list = event.target.files
    if not file_list.length:
        return

    file = file_list.item(0)

    # Read file content
    array_buffer = await file.arrayBuffer()
    bytes_data = bytes(Uint8Array.new(array_buffer))

    try:
        # Try to read the CSV with automatic delimiter detection
        text_data = bytes_data.decode("utf-8", errors="replace")
        sample = text_data[:4096]
        try:
            dialect = csv.Sniffer().sniff(sample, delimiters=[",", ";", "\t", "|"])
        except csv.Error:
            dialect = csv.excel

        reader = csv.DictReader(io.StringIO(text_data), dialect=dialect)
        csv_data = [
            row for row in reader if any((v or "").strip() for v in row.values())
        ]

        # Validate the CSV data
        is_valid, message = validate_csv_data(csv_data)
        if not is_valid:
            show_error(f"CSV validation failed: {message}")
            csv_data = None
            return

        # Get unique card types (excluding 'back' if present)
        unique_types = sorted(
            {row["type"] for row in csv_data if row.get("type") != "back"}
        )

        # Generate color pickers for each type
        generate_card_type_colors(unique_types)

        show_success(
            f"CSV file loaded successfully! Found {len(csv_data)} cards with {len(unique_types)} unique type(s): {', '.join(sorted(unique_types))}."
        )

    except Exception as e:
        show_error(
            f"Error reading CSV file: {str(e)}. Please ensure your file is in CSV format with semicolon or comma separators."
        )
        # Reset the input value to allow re-upload
        document.getElementById("csv-file").value = ""
        # Reset the csv_data variable
        csv_data = None


def create_card_front(record, pdf, config):
    """Creates a square card with the record data."""
    CARD_SIZE = config["CARD_SIZE"]
    COLOR = config["CARD_COLORS"][record["type"]]

    card_x = pdf.get_x()
    card_y = pdf.get_y()
    pdf.set_fill_color(*COLOR["bg"])
    pdf.rect(pdf.get_x(), pdf.get_y(), CARD_SIZE, CARD_SIZE, style="F")
    pdf.set_fill_color(255, 255, 255)

    # Place the center in the center of the card
    if str(record.get("center", "")).strip():
        pdf.set_text_color(*COLOR["text"])
        pdf.set_font(size=48, style="B")
        pdf.set_xy(card_x + (CARD_SIZE / 2), card_y + CARD_SIZE / 2 - pdf.font_size / 2)
        pdf.cell(CARD_SIZE, None, str(record["center"]), align="X")

    # Place the top text above the center
    pdf.set_font(size=14, style="B")
    pdf.set_xy(card_x + (CARD_SIZE / 2), card_y + 6)
    pdf.multi_cell(CARD_SIZE - 4, None, record["top"], align="X")

    # Place the bottom text below the center
    if str(record.get("bottom", "")).strip():
        pdf.set_font(style="I")
        pdf.set_xy(card_x + (CARD_SIZE / 2), card_y + CARD_SIZE - 14)
        pdf.multi_cell(CARD_SIZE - 4, None, record["bottom"], align="X")


def create_card_back(record, pdf, config):
    """Creates a square card showing a QR code of the 'url' property of the record."""
    CARD_SIZE = config["CARD_SIZE"]
    COLOR = config["CARD_COLORS"]["back"]

    card_x = pdf.get_x()
    card_y = pdf.get_y()
    pdf.set_fill_color(*COLOR["bg"])
    pdf.rect(pdf.get_x(), pdf.get_y(), CARD_SIZE, CARD_SIZE, style="F")
    pdf.set_fill_color(0, 0, 0)

    qr = qrcode.QRCode(
        version=1,
        error_correction=qrcode.constants.ERROR_CORRECT_L,
        box_size=10,
        border=4,
    )

    qr.add_data(record["url"])
    qr.make(fit=True)
    img = qr.make_image(fill_color=COLOR["text"], back_color="transparent")

    # Convert QR code to bytes
    img_bytes = io.BytesIO()
    img.save(img_bytes, format="PNG")
    img_bytes.seek(0)

    # Add image to PDF
    pdf.image(img_bytes, card_x + 1, card_y + 1, CARD_SIZE - 1, CARD_SIZE - 1)


def create_pdf_bytes(data, config):
    """Creates a PDF with the data from the DataFrame and returns it as bytes."""
    PAPER_WIDTH = config["PAPER_WIDTH"]
    PAPER_HEIGHT = config["PAPER_HEIGHT"]
    BLEED = config["BLEED"]
    CARD_SIZE = config["CARD_SIZE"]
    GAP = config["GAP"]
    CARD_MARGIN = config["CARD_MARGIN"]

    pdf = FPDF(format=[PAPER_WIDTH + 2 * BLEED, PAPER_HEIGHT + 2 * BLEED], unit="mm")
    pdf.set_font("Helvetica", size=12)
    pdf.set_margins(
        BLEED + CARD_MARGIN,
        BLEED + CARD_MARGIN,
        BLEED + CARD_MARGIN,
    )

    # Calculate the number of columns and rows that fit on the page.
    n_columns = int((PAPER_WIDTH - 2 * CARD_MARGIN + GAP) // (CARD_SIZE + GAP))
    n_rows_per_page = int((PAPER_HEIGHT - 2 * CARD_MARGIN + GAP) // (CARD_SIZE + GAP))
    n_rows = math.ceil(len(data) / n_columns)

    # Calculate the number of pages needed.
    n_pages = math.ceil(n_rows / n_rows_per_page)

    for page in range(n_pages):
        # Get the cards for this page.
        start_idx = page * n_rows_per_page * n_columns
        end_idx = (page + 1) * n_rows_per_page * n_columns
        cards_on_page = data[start_idx:end_idx]
        pdf.add_page()

        # Create the front of the cards, starting at the top left corner of the page
        for i in range(n_columns):
            row_x = pdf.get_x()
            row_y = pdf.get_y()
            for j in range(n_rows_per_page):
                idx = i + j * n_columns
                if idx >= len(cards_on_page):
                    break
                record = cards_on_page[idx]
                # Set the position of the card
                pdf.set_xy(row_x, row_y + j * CARD_SIZE + j * GAP)
                create_card_front(record, pdf, config)
            # Set the position for the next row
            pdf.set_xy(row_x + CARD_SIZE + GAP, row_y)

        pdf.add_page()
        # Create the back of the cards, starting at the top right corner of the page
        # This is necessary for double-sided printing.
        for i in reversed(range(n_columns)):
            row_x = pdf.get_x()
            row_y = pdf.get_y()
            for j in range(n_rows_per_page):
                idx = i + j * n_columns
                if idx >= len(cards_on_page):
                    break
                record = cards_on_page[idx]
                # Set the position of the card
                pdf.set_xy(row_x, row_y + j * CARD_SIZE + j * GAP)
                create_card_back(record, pdf, config)
            # Set the position for the next row
            pdf.set_xy(row_x + CARD_SIZE + GAP, row_y)

    return pdf.output()


def collect_config():
    """Collect configuration from the form inputs."""
    global csv_data
    config = {}

    # Numeric configuration
    config["PAPER_WIDTH"] = float(document.getElementById("paper-width").value)
    config["PAPER_HEIGHT"] = float(document.getElementById("paper-height").value)
    config["BLEED"] = float(document.getElementById("bleed").value)
    config["CARD_SIZE"] = float(document.getElementById("card-size").value)
    config["GAP"] = float(document.getElementById("gap").value)
    config["CARD_MARGIN"] = float(document.getElementById("card-margin").value)

    # Color configuration - always include back
    config["CARD_COLORS"] = {
        "back": {
            "bg": rgb_from_hex(document.getElementById("back-bg").value),
            "text": rgb_from_hex(document.getElementById("back-text").value),
        },
    }

    # Dynamically collect colors for card types found in CSV
    if csv_data is not None:
        unique_types = {row["type"] for row in csv_data if row.get("type") != "back"}
        for card_type in sorted(unique_types):
            bg_input = document.getElementById(f"{card_type}-bg")
            text_input = document.getElementById(f"{card_type}-text")
            if bg_input and text_input:
                config["CARD_COLORS"][card_type] = {
                    "bg": rgb_from_hex(bg_input.value),
                    "text": rgb_from_hex(text_input.value),
                }

    return config


def generate_pdf(event=None):
    """Generate the PDF from the loaded CSV data and configuration."""
    global csv_data

    # Check if CSV data is loaded
    if csv_data is None:
        show_error("Please upload a CSV file first.")
        return

    # Collect configuration
    try:
        config = collect_config()
    except Exception as e:
        show_error(f"Error reading configuration: {str(e)}")
        return

    # Validate configuration
    is_valid, message = validate_config(config)
    if not is_valid:
        show_error(message)
        return

    # Generate PDF
    try:
        show_status("Generating PDF... This may take a moment.")

        # Disable button during generation
        btn = document.getElementById("generate-btn")
        btn.disabled = True

        pdf_bytes = create_pdf_bytes(csv_data, config)

        # Create a download link
        pdf_array = Uint8Array.new(len(pdf_bytes))
        for i, b in enumerate(pdf_bytes):
            pdf_array[i] = b

        blob = Blob.new([pdf_array], {type: "application/pdf"})
        url = URL.createObjectURL(blob)

        # Create a temporary link and click it to download
        a = document.createElement("a")
        a.href = url
        a.download = "cards.pdf"
        document.body.appendChild(a)
        a.click()
        document.body.removeChild(a)
        URL.revokeObjectURL(url)

        show_success(
            "PDF generated successfully! Your download should start automatically."
        )

        # Re-enable button
        btn.disabled = False

    except Exception as e:
        show_error(f"Error generating PDF: {str(e)}")
        # Re-enable button
        document.getElementById("generate-btn").disabled = False


# Set up event listeners
def setup():
    """Set up event listeners for the page."""
    file_input = document.getElementById("csv-file")
    file_input.addEventListener("change", create_proxy(load_csv_file))


# Make generatePDF available globally for the button onclick
from js import window

window.generatePDF = create_proxy(generate_pdf)

# Run setup when the script loads
setup()

File: web_pdf_generator.js
// Global variables to store configuration
let csvData = null;

// Palette for card categories (400 as background, 950 as text)
const COLOR_PALETTE = [
  { bg: "#FF6466", text: "#480808" },
  { bg: "#FF8805", text: "#451305" },
  { bg: "#FFB800", text: "#481901" },
  { bg: "#FFC600", text: "#451F05" },
  { bg: "#A7E400", text: "#1D2E04" },
  { bg: "#31DE74", text: "#062E16" },
  { bg: "#00D393", text: "#002C22" },
  { bg: "#00D4BE", text: "#002F2E" },
  { bg: "#00D3F2", text: "#003345" },
  { bg: "#00BDFF", text: "#002F4A" },
  { bg: "#00A4FF", text: "#002655" },
  { bg: "#608AFF", text: "#131C4C" },
  { bg: "#9687FF", text: "#201267" },
  { bg: "#B57EFF", text: "#330A66" },
  { bg: "#E570FF", text: "#48054F" },
  { bg: "#FB67B5", text: "#520523" },
  { bg: "#FF647D", text: "#4E0318" },
  { bg: "#8CA1B9", text: "#000617" },
];

function getPaletteColor(index) {
  return COLOR_PALETTE[index % COLOR_PALETTE.length];
}

/**
 * Convert hex color to RGB object
 */
function rgbFromHex(hexColor) {
  const hex = hexColor.replace("#", "");
  return {
    r: parseInt(hex.substring(0, 2), 16),
    g: parseInt(hex.substring(2, 4), 16),
    b: parseInt(hex.substring(4, 6), 16),
  };
}

/**
 * Display an error message to the user
 */
function showError(message) {
  window.showAlert(message, "error");
}

/**
 * Display a success message to the user
 */
function showSuccess(message) {
  window.showAlert(message, "success");
}

/**
 * Display a status message to the user
 */
function showStatus(message) {
  console.log(message);
  window.showAlert(message, "status");
}

function yieldToBrowser() {
  return new Promise((resolve) => setTimeout(resolve, 0));
}

/**
 * Generate color pickers for each unique card type found in the CSV
 */
function generateCardTypeColors(cardTypes) {
  const container = document.getElementById("dynamic-card-types");
  container.innerHTML = ""; // Clear existing color pickers

  // Hide the "no CSV" message
  const noCsvMsg = document.getElementById("no-csv-message");
  if (noCsvMsg) {
    noCsvMsg.classList.add("hidden");
  }

  // Generate color picker for each card type
  cardTypes.sort().forEach((cardType, index) => {
    const defaultColors = getPaletteColor(index);

    // Create the HTML structure for background color
    const bgDiv = document.createElement("div");

    const bgLabel = document.createElement("span");
    bgLabel.className = "font-medium text-slate-700 dark:text-slate-300";
    bgLabel.textContent = `${cardType.charAt(0).toUpperCase() + cardType.slice(1)} BG`;
    bgDiv.appendChild(bgLabel);

    const bgInput = document.createElement("input");
    bgInput.type = "color";
    bgInput.className =
      "ring-2 ring-slate-200 dark:ring-slate-600 rounded-full w-10 h-10 shrink-0";
    bgInput.id = `${cardType}-bg`;
    bgInput.value = defaultColors.bg;
    bgDiv.appendChild(bgInput);

    container.appendChild(bgDiv);

    // Create the HTML structure for text color
    const textDiv = document.createElement("div");

    const textLabel = document.createElement("span");
    textLabel.className = "font-medium text-slate-700 dark:text-slate-300";
    textLabel.textContent = `${cardType.charAt(0).toUpperCase() + cardType.slice(1)} Text`;
    textDiv.appendChild(textLabel);

    const textInput = document.createElement("input");
    textInput.type = "color";
    textInput.className =
      "ring-2 ring-slate-200 dark:ring-slate-600 rounded-full w-10 h-10 shrink-0";
    textInput.id = `${cardType}-text`;
    textInput.value = defaultColors.text;
    textDiv.appendChild(textInput);

    container.appendChild(textDiv);
  });
}

/**
 * Validate that the CSV data has the required columns and valid values
 */
function validateCsvData(data) {
  const requiredColumns = ["type", "url", "top", "center", "bottom"];

  if (!data || data.length === 0) {
    return { valid: false, message: "CSV data is empty" };
  }

  // Check if all required columns exist
  const missingColumns = requiredColumns.filter((col) => !(col in data[0]));
  if (missingColumns.length > 0) {
    return {
      valid: false,
      message: `Missing required columns: ${missingColumns.join(", ")}`,
    };
  }

  // Check required fields are not empty (center can be empty)
  for (let idx = 0; idx < data.length; idx++) {
    const row = data[idx];
    if (!String(row.type || "").trim()) {
      return {
        valid: false,
        message: `Type column cannot contain empty values (row ${idx + 1})`,
      };
    }
    if (!String(row.url || "").trim()) {
      return {
        valid: false,
        message: `URL column cannot contain empty values (row ${idx + 1})`,
      };
    }
    if (!String(row.top || "").trim()) {
      return {
        valid: false,
        message: `Top column cannot contain empty values (row ${idx + 1})`,
      };
    }
    if (!String(row.bottom || "").trim()) {
      return {
        valid: false,
        message: `Bottom column cannot contain empty values (row ${idx + 1})`,
      };
    }
  }

  return { valid: true, message: "CSV data is valid" };
}

/**
 * Reset the CSV upload section to initial state
 */
function resetCsvUpload() {
  // Reset the file input
  document.getElementById("csv-file").value = "";
  csvData = null;

  // Hide success state and show upload prompt
  document.getElementById("csv-upload-success").classList.add("hidden");
  document.getElementById("csv-upload-prompt").classList.remove("hidden");
}

/**
 * Update the upload UI to show the uploaded filename
 */
function updateUploadUI(filename) {
  document.getElementById("csv-filename").textContent = filename;
  document.getElementById("csv-upload-prompt").classList.add("hidden");
  document.getElementById("csv-upload-success").classList.remove("hidden");
}

/**
 * Load and validate the CSV file
 */
async function loadCsvFile(event) {
  const fileList = event.target.files;
  if (!fileList.length) {
    return;
  }

  const file = fileList[0];

  try {
    // Use PapaParse to parse the CSV file
    Papa.parse(file, {
      header: true,
      skipEmptyLines: true,
      complete: function (results) {
        // Filter out completely empty rows
        csvData = results.data.filter((row) =>
          Object.values(row).some((v) => (v || "").trim()),
        );

        // Validate the CSV data
        const validation = validateCsvData(csvData);
        if (!validation.valid) {
          showError(`CSV validation failed: ${validation.message}`);
          csvData = null;
          return;
        }

        // Get unique card types (excluding 'back' if present)
        const uniqueTypes = [
          ...new Set(
            csvData.map((row) => row.type).filter((type) => type !== "back"),
          ),
        ];

        // Update the upload UI with the filename
        updateUploadUI(file.name);

        // Generate color pickers for each type
        generateCardTypeColors(uniqueTypes);

        showSuccess(
          `CSV file loaded successfully! Found ${csvData.length} cards with ${uniqueTypes.length} unique type(s): ${uniqueTypes.sort().join(", ")}.`,
        );
      },
      error: function (error) {
        showError(
          `Error reading CSV file: ${error.message}. Please ensure your file is in CSV format.`,
        );
        // Reset the input value to allow re-upload
        resetCsvUpload();
      },
    });
  } catch (e) {
    showError(
      `Error reading CSV file: ${e.message}. Please ensure your file is in CSV format.`,
    );
    // Reset the input and UI
    resetCsvUpload();
  }
}

/**
 * Create the front of a card
 */
function createCardFront(record, pdf, config, x, y) {
  const CARD_SIZE = config.CARD_SIZE;
  const COLOR = config.CARD_COLORS[record.type];

  // Draw background
  const bgColor = rgbFromHex(COLOR.bg);
  pdf.setFillColor(bgColor.r, bgColor.g, bgColor.b);
  pdf.rect(x, y, CARD_SIZE, CARD_SIZE, "F");

  // Set text color
  const textColor = rgbFromHex(COLOR.text);
  pdf.setTextColor(textColor.r, textColor.g, textColor.b);

  // Place the center in the center of the card
  if (String(record.center || "").trim()) {
    const centerText = String(record.center);
    const maxWidth = CARD_SIZE - 8; // keep small side padding
    const baseSize = 48;
    const minSize = 14;

    pdf.setFont("helvetica", "bold");
    pdf.setFontSize(baseSize);

    const baseWidth = pdf.getTextWidth(centerText);
    const scale = Math.min(1, maxWidth / baseWidth);
    const finalSize = Math.max(minSize, Math.floor(baseSize * scale));

    pdf.setFontSize(finalSize);
    pdf.text(centerText, x + CARD_SIZE / 2, y + CARD_SIZE / 2, {
      align: "center",
      baseline: "middle",
    });
  }

  // Place the top text above the center
  pdf.setFontSize(14);
  pdf.setFont("helvetica", "bold");
  const topLines = pdf.splitTextToSize(record.top, CARD_SIZE - 4);
  // Calculate line height and adjust position to center multi-line text
  const topLineHeight = pdf.getLineHeight() / pdf.internal.scaleFactor;
  const topOffset = ((topLines.length - 1) * topLineHeight) / 2;
  pdf.text(topLines, x + CARD_SIZE / 2, y + 14 - topOffset, {
    align: "center",
  });

  // Place the bottom text below the center
  if (String(record.bottom || "").trim()) {
    pdf.setFont("helvetica", "italic");
    const bottomLines = pdf.splitTextToSize(record.bottom, CARD_SIZE - 4);
    // Calculate line height and adjust position to center multi-line text
    const bottomLineHeight = pdf.getLineHeight() / pdf.internal.scaleFactor;
    const bottomOffset = ((bottomLines.length - 1) * bottomLineHeight) / 2;
    pdf.text(
      bottomLines,
      x + CARD_SIZE / 2,
      y + CARD_SIZE - 10 - bottomOffset,
      {
        align: "center",
      },
    );
  }
}

/**
 * Create the back of a card with QR code
 */
async function createCardBack(record, pdf, config, x, y) {
  const CARD_SIZE = config.CARD_SIZE;
  const COLOR = config.CARD_COLORS.back;

  // Draw background
  const bgColor = rgbFromHex(COLOR.bg);
  pdf.setFillColor(bgColor.r, bgColor.g, bgColor.b);
  pdf.rect(x, y, CARD_SIZE, CARD_SIZE, "F");

  // Generate QR code using QRious
  const textColor = rgbFromHex(COLOR.text);
  const qr = new QRious({
    value: record.url,
    size: 500, // High resolution for quality
    foreground: `rgb(${textColor.r}, ${textColor.g}, ${textColor.b})`,
    background: `rgb(${bgColor.r}, ${bgColor.g}, ${bgColor.b})`,
    level: "L",
  });

  // Add QR code to PDF
  const qrDataUrl = qr.toDataURL("image/jpeg", 0.85);
  pdf.addImage(qrDataUrl, "JPEG", x + 8, y + 8, CARD_SIZE - 16, CARD_SIZE - 16);
}

/**
 * Create PDF with all cards
 */
async function createPdfBytes(data, config) {
  const PAPER_WIDTH = config.PAPER_WIDTH;
  const PAPER_HEIGHT = config.PAPER_HEIGHT;
  const BLEED = config.BLEED;
  const CARD_SIZE = config.CARD_SIZE;
  const GAP = config.GAP;
  const CARD_MARGIN = config.CARD_MARGIN;

  // Create PDF with custom page size
  const { jsPDF } = window.jspdf;
  const pdf = new jsPDF({
    orientation: "portrait",
    unit: "mm",
    format: [PAPER_WIDTH + 2 * BLEED, PAPER_HEIGHT + 2 * BLEED],
  });

  pdf.setFont("helvetica");

  // Calculate the number of columns and rows that fit on the page
  const nColumns = Math.floor(
    (PAPER_WIDTH - 2 * CARD_MARGIN + GAP) / (CARD_SIZE + GAP),
  );
  const nRowsPerPage = Math.floor(
    (PAPER_HEIGHT - 2 * CARD_MARGIN + GAP) / (CARD_SIZE + GAP),
  );
  const nRows = Math.ceil(data.length / nColumns);

  // Calculate the number of pages needed
  const nPages = Math.ceil(nRows / nRowsPerPage);

  for (let page = 0; page < nPages; page++) {
    // Get the cards for this page
    const startIdx = page * nRowsPerPage * nColumns;
    const endIdx = (page + 1) * nRowsPerPage * nColumns;
    const cardsOnPage = data.slice(startIdx, endIdx);

    // Add page for fronts
    if (page > 0) pdf.addPage();

    // Create the front of the cards, starting at the top left corner of the page
    for (let i = 0; i < nColumns; i++) {
      for (let j = 0; j < nRowsPerPage; j++) {
        const idx = i + j * nColumns;
        if (idx >= cardsOnPage.length) break;

        const record = cardsOnPage[idx];
        const cardX = BLEED + CARD_MARGIN + i * (CARD_SIZE + GAP);
        const cardY = BLEED + CARD_MARGIN + j * (CARD_SIZE + GAP);
        createCardFront(record, pdf, config, cardX, cardY);
      }
    }

    // Add page for backs
    pdf.addPage();

    // Create the back of the cards, starting at the top right corner of the page
    // This is necessary for double-sided printing
    for (let i = nColumns - 1; i >= 0; i--) {
      for (let j = 0; j < nRowsPerPage; j++) {
        const idx = i + j * nColumns;
        if (idx >= cardsOnPage.length) break;

        const record = cardsOnPage[idx];
        const cardX =
          BLEED + CARD_MARGIN + (nColumns - 1 - i) * (CARD_SIZE + GAP);
        const cardY = BLEED + CARD_MARGIN + j * (CARD_SIZE + GAP);
        await createCardBack(record, pdf, config, cardX, cardY);
      }
    }
  }

  return pdf;
}

/**
 * Collect configuration from the form inputs
 */
function collectConfig() {
  const config = {};

  // Numeric configuration
  config.PAPER_WIDTH = parseFloat(document.getElementById("paper-width").value);
  config.PAPER_HEIGHT = parseFloat(
    document.getElementById("paper-height").value,
  );
  config.BLEED = parseFloat(document.getElementById("bleed").value);
  config.CARD_SIZE = parseFloat(document.getElementById("card-size").value);
  config.GAP = parseFloat(document.getElementById("gap").value);
  config.CARD_MARGIN = parseFloat(document.getElementById("card-margin").value);

  // Color configuration - always include back
  config.CARD_COLORS = {
    back: {
      bg: document.getElementById("back-bg").value,
      text: document.getElementById("back-text").value,
    },
  };

  // Dynamically collect colors for card types found in CSV
  if (csvData !== null) {
    const uniqueTypes = [
      ...new Set(
        csvData.map((row) => row.type).filter((type) => type !== "back"),
      ),
    ];
    uniqueTypes.forEach((cardType) => {
      const bgInput = document.getElementById(`${cardType}-bg`);
      const textInput = document.getElementById(`${cardType}-text`);
      if (bgInput && textInput) {
        config.CARD_COLORS[cardType] = {
          bg: bgInput.value,
          text: textInput.value,
        };
      }
    });
  }

  return config;
}

/**
 * Generate the PDF from the loaded CSV data and configuration
 */
async function generatePDF(event) {
  // Check if CSV data is loaded
  if (csvData === null) {
    showError("Please upload a CSV file first.");
    return;
  }

  // Collect configuration
  let config;
  try {
    config = collectConfig();
  } catch (e) {
    showError(`Error reading configuration: ${e.message}`);
    return;
  }

  // Generate PDF
  try {
    // Disable button during generation
    const btn = document.getElementById("generate-btn");
    btn.disabled = true;

    showStatus("Generating PDF... This may take a moment.");
    await yieldToBrowser();

    const pdf = await createPdfBytes(csvData, config);

    // Save the PDF
    pdf.save("cards.pdf");

    showSuccess(
      "PDF generated successfully! Your download should start automatically.",
    );

    // Re-enable button
    btn.disabled = false;
  } catch (e) {
    showError(`Error generating PDF: ${e.message}`);
    console.error("PDF generation error:", e);
    // Re-enable button
    document.getElementById("generate-btn").disabled = false;
  }
}

// Set up event listeners when DOM is ready
document.addEventListener("DOMContentLoaded", function () {
  const fileInput = document.getElementById("csv-file");
  if (fileInput) {
    fileInput.addEventListener("change", loadCsvFile);
  }

  // Make generatePDF available globally
  window.generatePDF = generatePDF;
});
